### PR TITLE
Update (or save) a testcase. [PUT] /api/problems/{problemId}/testcase…

### PR DESCRIPTION
**Pull request by ShawnSWu**

### CODE REVIEW

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 1]
```diff
@@ -398,6 +394,54 @@ private void archiveOrDeleteProblem(int problemId) throws Exception {
...
+        Problem problem = saveProblemAndGet(problemId);
+
+        Testcase expectedTestcase = problem.getTestcases().get(0);
+        String testCaseName = expectedTestcase.getName();
+        expectedTestcase.setGrade(100);
+        updateOrAddTestCase(problemId, testCaseName, expectedTestcase);
+
+        var actualProblem = getProblem(problemId);
+        List<Testcase> testcases = actualProblem.getTestcases();
+        Testcase actualTestCase = findFirst(testcases, tc -> 100 == tc.getGrade()).orElseThrow();
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:409]:
```Java
        expectedTestcaseGrade = 100;
        expectedTestcase.setGrade(expectedTestcaseGrade);
        updateOrAddTestCase(problemId, testCaseName, expectedTestcase);

        var actualProblem = getProblem(problemId);
        List<Testcase> testcases = actualProblem.getTestcases();
        Testcase actualTestCase = findFirst(testcases, tc -> expectedTestcaseGrade == tc.getGrade()).orElseThrow();
```

File: Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java [hunk 2]
```diff
@@ -398,6 +394,54 @@ private void archiveOrDeleteProblem(int problemId) throws Exception {
...
+        saveProblemAndGet(problemId);
+
+        Testcase expectedTestcase = new Testcase("123456", problemId, 100, 300, 300, -100, 500);
+        updateOrAddTestCase(problemId, "123456", expectedTestcase);
+
+        List<Testcase> testcases = getProblem(problemId).getTestcases();
+        assertEquals(4, testcases.size());
+        Testcase actualTestcase = findFirst(testcases, testcase -> "123456".equals(testcase.getName())).orElseThrow();
+        assertNotNull(actualTestcase.getId());
+        assertTestCaseEquals(expectedTestcase, actualTestcase);
```
Comment (Wally5077) [inline Spring-Boot/Spring-Boot-Problem/src/test/java/tw/waterball/judgegirl/springboot/problem/controllers/ProblemControllerTest.java:425]:
```Java
        String expectedTestCaseName = "123456";
        Testcase expectedTestcase = new Testcase(expectedTestCaseName, problemId, 100, 300, 300, -100, 500);
        updateOrAddTestCase(problemId, expectedTestCaseName, expectedTestcase);

        List<Testcase> testcases = getProblem(problemId).getTestcases();
        assertEquals(4, testcases.size());
        Testcase actualTestcase = findFirst(testcases, testcase -> expectedTestCaseName.equals(testcase.getName())).orElseThrow();
        assertNotNull(actualTestcase.getId());
        assertTestCaseEquals(expectedTestcase, actualTestcase);
```

**Review by Wally5077 — CHANGES_REQUESTED** [review]
Review done

**Review by Johnny850807 — APPROVED** [review]
LGTM

**Review by timt1028 — APPROVED** [review]
LGTM

**Review by Wally5077 — APPROVED** [review]
LGTM

**Pull request merged**